### ISSUE TITLE
Light flare from lamps has incorrect irregular depth

### ISSUE THREAD

**Issue by TiberiumFusion** [Facepunch/garrysmod-issues#4138]
This is a bit tricky to describe in words, so please take a look at this demonstration video:
https://drive.google.com/open?id=1F34XgZz6IJcAeyOMQDiihx6y4InKL-1B

In short, the light flare on lamps has a much greater apparent depth than it should, when viewed from as little as a 10° angle. This behavior is irregular in that some props are affected, while other are not. The greater the distance from the lamp (and thus its light flare), the worse the problem gets, i.e. the less of a viewing angle you need to experience this problem. Also note from my demo video how the issue persists even when the lamp is at least 5 meters away from the background props (that are still somehow occluding its light flare).

The props in the video don't have any funny properties applied to them. They are all standard Builder or SProps (https://steamcommunity.com/workshop/filedetails/?id=173482196) props. Both standard Builder props and SProps are present in the demo video. All props have typical render properties and have default gmod materials that are (presumably) using VertexLitGeneric. I have the color of most props set to pure black, but this issue occurs regardless of prop color, as shown later in the video.

Sometimes, newly-spawned props can "correct" the issue when placed between the lamp and the pre-existing props which are occluding the light flare. These newly spawned props are correctly stenciled behind the light flare (or, the light flare is correctly stenciled above the new props). Getting this to work seems to be random, and I could not capture it on video when I wanted to. Watching the pot boil, I suppose. This may suggest, however, that the issue may be that random props are writing bad values to some depth stencil, instead of an issue with the lamp's light flare itself.

**I have encountered this in the x64 branch of gmod circa 2020-01-17. I haven't tested this the regular old version.**
My graphics hardware:
- HD 7970
- Driver package version 16.50.2011-161223a-310015C
- D3D (11.0) driver version 9.14.10.01235

**Comment by robotboy655** [comment]
Not gonna fix, this is how sprites work.

**Comment by TiberiumFusion** [comment]
@robotboy655 Alright. If you've got a spare moment, could you point me towards a description of sprite behavior? (if such exists) Since it won't be fixed, I'd like to make some sense of the irregularity.

**Comment by robotboy655** [comment]
It's a square that is always parallel to your screen. So it will go through walls/props depending on the angle your view is.

**Comment by TiberiumFusion** [comment]
@robotboy655 Certainly, if the billboard exists as a typical quad in worldspace, where the actual size of the geometry equals the apparent size of the billboard in the camera's view (and thus becomes subject to clipping). However, due to the irregularity (more on that below), I'm not sure if that's how it really works.

The worldspace billbaord strategy is ill-suited for the light flare effect, since the flare is a simulation of the starbust effect that occurs in our eyes. As such, it makes much more sense to define the light flare to a be a single point in worldspace - a "binary" point that can stencil its depth and will only ever be fully in view or fully occluded. The light flare texture is then simply rendered central to the point, with zero depth, in front of everything else. Thus, the light flare can exist in apparent screenspace and will not be unnaturally occluded by worldspace geometry, which is much more analogous to the how the real life phenomenon works. Surely the Source developers were aware of this?

And there's a couple ways to achieve it. I'm aware of Source's age, which most likely prevents primarily hardware-side implementations, but I can't imagine that one CPU ray trace per lamp (to check occlusion on the lamp's position) would be unacceptably expensive. In fact, the software ray trace strategy seems right up Source's alley, considering how many games of the early 00s used this. This wouldn't work on props without collision hulls, of course, but that seems like an acceptable loss, especially considering that gmod is a physics sandbox and that most props have collisions.

What is puzzling me is the irregularity in the light flare's clipping (that I showed in the demo video). Sometimes, I have managed to correct the light flare's clipping by creating a new prop and placing it in the **exact** location of a pre-existing prop which was suffering clipping by the light flare. Same prop, same materials, same size/dimensions. The new prop does not suffer any clipping with the light flare whatsoever (the light flare correctly occludes the entire prop), testing up to viewing angles of 45°. This suggests that the rendering strategy for the light flares is not a worldspace quad with depth writing. That is the irregularity that I'm trying to describe and understand.

It really feels like a glitch, in all honesty. There is another prop build I have which randomly exhibits the unnatural clipping when I paste it in, i.e. sometimes the light flare is always in front of all props (as it should be), while other times it suffers the kind of clipping I showed in my demo video.

**Comment by robotboy655** [comment]
Actually I am wrong about "this is how sprites work", I forget that the lamp overlay is drawn above everything else no matter what,

It is possible the glitches you are experiencing are a problem with the materials on the props behind the lamp. Have you tried using built-in hl2 props? It could also be the materials you are applying.
